Pd 807 nf sv 4 smb mixed mode shares (#2246)
truenas/documentation: content/SCALE/SCALETutorials/Shares/MixedModeShares.md

---
title: "Multiprotocol Shares"
description: "Provides instructions on setting up SMB and NFSv4 mixed-mode shares."
weight: 40
tags:
- scaleshares
- scalesmb
- scalenfs
---

{{< toc >}}

## About Multiprotocol Shares

A multiprotocol or mixed-mode NFS and SMB share supports both NFS and SMB protocols for sharing data.

> Multiprotocol shares allow clients running different operating systems to access the same data.

Reviewer: allows clients to use different file sharing protocols

> This can be particularly useful in environments with a mix of Unix-like systems (which prefer NFS) and Windows systems (which prefer SMB).

Reviewer: We need to emphasize that the preferred configuration is to use the Linux SMB client rather than a multiprotocol share.

It is important to properly configure permissions and access controls to ensure security and data integrity when using mixed-mode sharing.
To maximize security on the NFS side of the multiprotocol share, we recommend using NFSv4 and [Active Directory](#joining-active-directory) for Kerberos authentication.
It is also important that NFS clients preserve extended attributes when copying files, or SMB metadata could be discarded in the copy.

## Adding a Multiprotocol Share

Adding a multiprotocol SMB and NFS share to your system involves several steps to add the share and get it working.

1. [Configure Active Directory](#joining-active-directory) and join the TrueNAS SCALE system to your AD domain.

>    Configure a container, Kerberos admin, and user accounts in AD.

Reviewer: You can just simplify to "Join the TrueNAS server to an existing Active Directory domain."

2. [Set up a dataset](#creating-a-multiprotocol-share-dataset) for the new share with **Share Type** set to **Multiprotocol**.

3. [Create the SMB share](#creating-the-smb-share) with **Purpose** set to **Multi-protocol (NFSv4/SMB) shares**.

> 4. [Create the NFS share](#creating-the-nfs-share) with **Security** set to **KRB5**.

Reviewer: There is a step involving NFS configuration where we have a button in the NFS form to add the NFS kerberos (KRB5) service principle name (SPN) to the AD machine account for the TrueNAS server in Active Directory. If KRB5 SPN is not present in local kerberos keytab and in AD machine account, then kerberized NFS clients will not be able to mount the NFS export.

Author: To address this I'm moving this step to the beginning and will include a tip that you will need to complete the form if AD is already joined.

5. Start and [configure the SMB and NFS services](#configuring-the-smb-and-nfs-services) to begin sharing.

6. [Connect client system(s)](#connecting-to-a-multiprotocol-share) to the share.

### Joining Active Directory

Mixed-mode SMB and NFS shares greatly simplify data access for client running a range of operating systems.
They also require careful attention to security complexities not present in standard SMB shares.

> NFS shares do not evaluate permissions set in the SMB Share ACL.

Reviewer: s/evaluate/respect maybe link to docs explaining SMB share ACL vs filesystem ACL.

{{< trueimage src="/images/SCALE/Shares/NFSWidgetOptions.png" alt="Unix (NFS) Shares Widget Options" id="Unix (NFS) Shares Widget Options" >}}

Each NFS share on the list also has a toggle to enable or disable the service for that share.

#### Starting the Service Using System Settings

Go to **System Settings > Services** and click the toggle for **NFS**.
Set **Start Automatically** if you want the service to activate when TrueNAS boots.

{{< hint type=note >}}
The NFS service does not automatically start on boot if all NFS shares are encrypted and locked.
{{< /hint >}}
{{< /expand >}}

### Configuring the NFS Service

Configure the NFS service by clicking **Config Service** from the <span class="material-icons">more_vert</span> dropdown menu on the **Shares** screen or by clicking <i class="material-icons" aria-hidden="true" title="Configure">edit</i> on the **Services** screen.

Under **NFSv4**, ensure the **NFSv4** protocol is selected from the **Enabled Protocols** dropdown menu.
For security hardening, we recommend disabling the **NFSv3** protocol.

> Select **Require Kerberos for NFSv4** to enable the Kerberos ticket that generated when you [joined active directory](#joining-active-directory).

Reviewer: We probably need to add steps regarding how to enable this properly if it involves adding KRB5 SPN as mentioned above.

Multiprotocol shares allow clients to use either protocol to access the same data.
This can be useful in environments with a mix of Windows systems and Unix-like systems, especially if some clients lack an SMB client.

{{< hint type=tip >}}
Carefully consider your environment and access requirements before configuring a multiprotocol share.
For many applications, a single protocol SMB share provides better user experience and ease of administration. Linux clients can access SMB shares using [`mount.cifs`](https://linux.die.net/man/8/mount.cifs).
{{< /hint >}}

It is important to properly configure permissions and access controls to ensure security and data integrity when using mixed-mode sharing.
To maximize security on the NFS side of the multiprotocol share, we recommend using NFSv4 and [Active Directory](#joining-active-directory) for Kerberos authentication.
It is also important that NFS clients preserve extended attributes when copying files, or SMB metadata could be discarded in the copy.

## Adding a Multiprotocol Share

Adding a multiprotocol SMB and NFS share to your system involves several steps:

> 1. [Start and configure](#starting-and-configuring-services) the SMB and NFS services.

Reviewer 2: ```suggestion
1. [Configure and start](#starting-and-configuring-services) the SMB and NFS services.
```

2. Join the TrueNAS server to an existing [Active Directory](#joining-active-directory) domain.
   Configure a container, Kerberos admin, and user accounts in AD.

> 3. [Set up a dataset](#creating-a-multiprotocol-share-dataset) for the new share with **Share Type** set to **Multiprotocol**.

Reviewer 2: In the order of configuring shares, you first create the dataset, then the share so this is a preparation step or should come earlier in this overview procedure. If you create a preparation or first steps section, then you can mention this in the procedure intro, as in "After you take care of the first steps (link to it), then: and then list the steps for the overview. Leverage links to existing instructions like "AddSMBShare.md" and "AddingNFSShares.md" for the details. If only part of the instructions apply in this procedure, link to the sections that apply.

Author: The order is different here because you first need to configure the NFS service to work with KBR5, then join AD, then proceed with the typical order of creating the dataset and creating the shares.
 
The content here does leverage a lot from the existing articles, but the settings are unique enough that it isn't 1 to 1, plus we want to avoid bouncing people from page to page too much and make this a cohesive tutorial. The bottom section does link over because there are no unique instructions for connecting client systems once the share is set up.

Author: We could  look into snippetizing some sections to make updates easier, but this really needs to merge today (and be ported/merged to 23.10), so that might be an improvement to look at later this week.

Reviewer 2: I think we can create a ticket to do the snippetizing and not worry about it here.

> 4. [Create the SMB share](#creating-the-smb-share) with **Purpose** set to **Multi-protocol (NFSv4/SMB) shares**.

Reviewer 2: can you link to the existing "AddSMBShare.md" article here and get rid of the extra section that doesn't really add any additional value?

> 5. [Create the NFS share](#creating-the-nfs-share) with **Security** set to **KRB5**.

Reviewer 2: Same comment as the SMB share step. Can you link to the "AddingNFSShares.md" article?

6. [Connect client system(s)](#connecting-to-a-multiprotocol-share) to the share.

> ## Starting and Configuring Services

Author: I've added this to the beginning of the section

> Before joining AD and creating a dataset for the share to use, first start both the SMB and NFS services and configure the NFS Service for Kerberos authentication. Configure the NFS service before joining AD for simpler Kerberos credential creation.
 
 These are the steps I was told to document, so they do need to be here.

{{< expand "Creating a Dataset" "v" >}}
{{< include file="/content/_includes/CreateDatasetSCALE.md" >}}
{{< /expand >}}

#### Adjusting the Dataset ACL

After joining AD and creating a dataset, you need to adjust the dataset/filesystem ACL to match the container and users configured in AD.

{{< expand "Click here for instructions" "v" >}}
1. Go to **Datasets**.

2. Click on the name of the dataset you created for the multiprotocol share to use.

3. Scroll down to the **Permissions** widget. Click **Edit** to open the **Edit ACL** screen.

4. Check the **Access Control List** to see if the AD group you created is on the list and has the correct permissions. If not, add this Access Control Entry (ACE) item.

   a. Enter **Group** in the **Who** field or use the dropdown list to select **Group**.

   b. Type or select the appropriate group in the **Group** field.

   c. Verify **Full Control** displays in **Permissions**. If not, select it from the dropdown list.

   d. Click **Save Access Control List** to add the ACE item or save changes.

See [Permissions]({{< relref "PermissionsScale.md" >}}) for more information on editing dataset permissions.
{{< /expand >}}

> ### Creating the SMB Share

Reviewer 2: is there a reason why you can't link to the existing procedure instead of rewriting the instructions here? If the only thing that differs is the dataset you select, you can say this earlier in the overview procedure.

Author: Again, I think there's value in having a single cohesive procedure here rather than linking out. Every step has some unique variable and I think as a reader I'd rather have a continuous process than constantly have to read another article but also remember the specific ways in which my application differs from the instructions in that other article.

   If you do not enter a name, the share name becomes the last component of the path.
   If you change the name, follow the naming conventions for:
   * [Files and directories](https://learn.microsoft.com/en-us/windows/win32/fileio/naming-a-file#naming-conventions)
   * [Share names](https://learn.microsoft.com/en-us/openspecs/windows_protocols/ms-fscc/dc9978d7-6299-4c5a-a22d-a039cdc716ea)

3. Select **Multi-protocol (NFSv4/SMB) shares** from the **Purpose** dropdown list to apply pre-determined **Advanced Options** settings for the share.

4. (Optional) Enter a **Description** to help explain the share purpose.

5. Select **Enabled** to allow sharing of this path when the SMB service is activated.
   Leave it cleared if you want to disable the share without deleting the configuration.

6. If needed, use **Advanced Options** to set up guest access, read only access, to set up allowed and denied hosts. or to optimize the SMB share for Apple OS.
   See [Adding SMB Shares]({{< relref "AddSMBShares.md #configuring-share-advanced-options-settings" >}}) for more information.

7. Click **Save** to create the share and add it to the **Shares > Windows (SMB) Shares** list.

> ### Creating the NFS Share

Reviewer 2: same comments here as in SMB share. If the current instructions cover this with the only change being the name of the dataset to use or a permission/setting option, then try to use a link to that procedure and only call out the difference.

Enter an IP address in **Network** and select the mask CIDR notation.
Click **Add** for each network address and CIDR you want to define as an authorized network.
Defining an authorized network restricts access to all other networks. Leave empty to allow all networks.

If you want to enter allowed systems, click **Add** to the right of **Hosts**.
Enter a host name or IP address to allow that system access to the NFS share.
Click **Add** for each allowed system you want to define.
Defining authorized systems restricts access to all other systems.
Press the **X** to delete the field and allow all systems access to the share.
    {{< /expand >}}

5. Enable Kereberos security.

    Click **Advanced Options**.

    {{< trueimage src="/images/SCALE/Shares/AddNFSAdvancedOptionsAccessSettings.png" alt="Advanced Options Access Settings" id="Advanced Options Access Settings" >}}

    If needed, select **Read-Only** to prohibit writing to the share.

    Select **KRB5** from the **Security** dropdown to enable the Kerberos ticket that generated when you [joined Active Directory](#joining-active-directory).

6. Click **Save** to create the share.

## Connecting to a Multiprotocol Share

> Once created and configured, you can connect to your mulitprotocol share using either SMB or NFS and from a variety of client operating systems including Windows, Apple, FreeBSD, and Linux/Unix systems.

Reviewer 2: After, not once
```suggestion
After you create and configure the shares, connect to your mulitprotocol share using either SMB or NFS or from a variety of client operating systems including Windows, Apple, FreeBSD, and Linux/Unix systems.
```
